vermouth/gmx/topology.py

```diff
@@ -0,0 +1,216 @@
+"""
+I/O of topology parameters that are not molecules.
+"""
+import itertools
+from collections import namedtuple
+import textwrap
+import vermouth
+from vermouth.file_writer import deferred_open
+from vermouth.citation_parser import citation_formatter
+from ..log_helpers import StyleAdapter, get_logger
+from .itp import _interaction_sorting_key
+
+LOGGER = StyleAdapter(get_logger(__name__))
+
+Atomtype = namedtuple('Atomtype', 'molecule node sigma epsilon meta')
+NonbondParam = namedtuple('NonbondParam', 'atoms sigma epsilon meta')
+
+def _group_by_conditionals(interactions):
+    interactions_group_sorted = sorted(interactions,
+        key=_interaction_sorting_key
+        )
+    interaction_grouped = itertools.groupby(
+        interactions_group_sorted,
+        key=_interaction_sorting_key
+        )
+    return interaction_grouped
+
+def convert_sigma_epsilon(sigma, epsilon):
+    """
+    Convert the LJ potential from sigma epsilon
+    form to C6 C12 form.
+    """
+    C6 = 4*sigma*epsilon**6
+    C12 = 4*sigma*epsilon**12
+    return C6, C12
+
+def write_atomtypes(system, itp_path, C6C12=False):
+    """
+    Writes the [atomtypes] directive to file.
+    All atomtypes are defined in system.gmx_topology_params.
+    Masses and further information are taken from the molecule
+    directly.
+    """
+    conditional_keys = {True: '#ifdef', False: '#ifndef'}
+    with deferred_open(itp_path, "w") as itp_file:
+        itp_file.write("[ atomtypes ]\n")
+        grouped_types = _group_by_conditionals(system.gmx_topology_params['atomtypes'])
+        for (conditional, group), interactions_in_group in grouped_types:
+            if conditional:
+                conditional_key = conditional_keys[conditional[1]]
+                itp_file.write('{} {}\n'.format(conditional_key, conditional[0]))
+            if group:
+                itp_file.write('; {}\n'.format(group))
```

Comment:
I would appreciate some comments here I think

Reply:
me too. this comes straight from the moleule itp writer and is black magic; I put a comment that points this out